apps/arena/native/physics/src/collision_detection/ear_clipping.rs

```diff
@@ -0,0 +1,203 @@
+use crate::map::{Entity, Position};
+
+/*
+Ear clipping triangulation algorithm
+
+To summarize the algorithm works by traversing the vertices from a polygon finding ears an deleting the current vertex
+from the polygon shape and repeat until there's 3 vertices in the polygon
+
+An ear is defined by the following:
+ a. It's a triangle formed by the current vertex we're traversing, the previous one and the next
+ b. The inner angle formed by the three vertices aren't greater than 180 degrees
+ c. There isn't any vertex inside the triangle area
+
+The logic of the algorithm goes like this:
+1. Traverse the polygon looking for ears
+2. Once we find an ear do the following:
+   - Save the triangle formed by the current previous and next vertex
+   - Delete the current vertex from the available vertex list
+   - Start again with the cleaned up vertex list until the list length is 3
+3. Make a final triangle with the 3 remaining vertices
+
+Even though the resolution is different, the logic is the same as the following article:
+- [Geometric tools article](https://www.geometrictools.com/Documentation/TriangulationByEarClipping.pdf)
+
+*/
+
+pub(crate) fn maybe_triangulate_polygon(mut polygon: Entity) -> Vec<Entity> {
+    remove_redundant_vertices(&mut polygon);
+    if is_polygon_convex(&polygon) {
+        return vec![polygon.clone()];
+    }
+    triangulate_polygon(&polygon)
+}
+
+// Remove redundant vertices.
+// A redundant vertex is one that is in the middle of a straight line drawn by two other vertices.
+fn remove_redundant_vertices(polygon: &mut Entity) {
+    let mut cleaned_vertices = vec![];
+    for current_vertex_index in 0..polygon.vertices.len() {
+        let previous_vertex_index =
+            get_previous_vertex_index(current_vertex_index, &polygon.vertices);
+        let next_vertex_index = get_next_vertex_index(current_vertex_index, &polygon.vertices);
+
+        let previous_vertex = polygon.vertices[previous_vertex_index];
+        let current_vertex = polygon.vertices[current_vertex_index];
+        let next_vertex = polygon.vertices[next_vertex_index];
+
+        if !is_vertex_consecutive(&previous_vertex, &current_vertex, &next_vertex) {
+            cleaned_vertices.push(current_vertex);
+        }
+    }
+
+    polygon.vertices = cleaned_vertices;
+}
+
+fn is_vertex_consecutive(
+    previous_vertex: &Position,
+    current_vertex: &Position,
+    next_vertex: &Position,
+) -> bool {
+    let current_to_previous_vector = Position::sub(previous_vertex, current_vertex);
+    let current_to_next_vector = Position::sub(next_vertex, current_vertex);
+
+    get_cross_product_value(&current_to_previous_vector, &current_to_next_vector) == 0.0
+}
+
+fn triangulate_polygon(polygon: &Entity) -> Vec<Entity> {
+    let mut result: Vec<Entity> = vec![];
+    let mut vertices = polygon.vertices.clone();
+    let mut ear_found: bool;
+
+    while vertices.len() > 3 {
```

Comment:
This while loop determines whether we can find a ear triangle or not by given some vertices right? Can we create an auxiliary function for this?

Reply:
Thanks for the explanation!